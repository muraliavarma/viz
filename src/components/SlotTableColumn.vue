<script>
import SlotTableHeader from './SlotTableHeader.vue';

export default {
  name: 'SlotTableColumn',
  functional: true,
  props: {
    sticky: {
      type: String,
      default: '',
    },
  },
  render(createElement, ctx) {
    const { sticky } = ctx.props;
    const data = Object.assign({}, ctx.data);
    if (sticky === 'left') {
      data.class = Object.assign({}, data.class, { 'sticky-left': true });
    } else if (sticky === 'right') {
      data.class = Object.assign({}, data.class, { 'sticky-right': true });
    }

    // header can be rendered just with the column info, so simply render it in its own component
    const header = createElement(
      SlotTableHeader,
      data,
      ctx.children.filter(child => child.data.slot === 'header'),
    );

    // cells are more complicated, since it needs row info as well
    // so we simply return a scoped slot that will be rendered eventually on a per row basis
    // if, instead, the consumer passed a simple non-scoped slot, we send back a simple slot
    let cell;
    if (ctx.data.scopedSlots && ctx.data.scopedSlots.cell) {
      ({ cell } = ctx.data.scopedSlots);
      delete ctx.data.scopedSlots.cell;
    } else if (ctx.children.length > 0) {
      cell = ctx.children.find(child => child.data.slot === 'cell');
    }

    // send back the header and cell data as props
    const props = Object.assign({}, data.props, {
      header, cell,
    });
    return createElement('div', Object.assign(
      {},
      data, { props },
    ));
  },
};
</script>
