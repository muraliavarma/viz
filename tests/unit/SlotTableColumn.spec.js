import { mount } from '@vue/test-utils';
import SlotTableColumn from '@/components/SlotTableColumn.vue';

const headerSlot = {
  data: {
    slot: 'header',
    foo: 'bar',
  },
};

const cellSlot = {
  data: {
    slot: 'cell',
    baz: 'jaa',
  },
};

describe('SlotTableColumn.vue', () => {
  it('has the expected html structure', () => {
    const wrapper = mount(SlotTableColumn, {
      context: {
        children: [headerSlot, cellSlot],
        staticClass: 'bar',
        attrs: {
          baz: 'quux',
        },
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('passes the header and cell slots down', () => {
    const wrapper = mount(SlotTableColumn, {
      context: {
        children: [headerSlot, cellSlot],
      },
    });
    expect(wrapper.vnode.data.props.header.children[0]).toBe(headerSlot);
    expect(wrapper.vnode.data.props.cell).toBe(cellSlot);
  });

  it('passes the cell scoped slot down', () => {
    const cellTemplate = '<div slot="cell" slot-scope="{ row }">{{row + \'-\' + row}}</div>';
    const wrapper = mount(SlotTableColumn, {
      scopedSlots: { cell: cellTemplate },
      context: {
        props: {
          foo: 'bar',
        },
      },
    });
    // remove the scoped slot from parent
    expect(wrapper.vnode.data.scopedSlots.cell).toBe(undefined);
    // pass in any additional prop into child
    expect(wrapper.vnode.data.props.foo).toBe('bar');
    // verify that the cell slot scope got passed down
    expect(wrapper.vnode.data.props.cell({ row: 'foo' }).children[0].text).toBe('foo-foo');
  });

  it('does left sticky', () => {
    const stickyWrapper = mount(SlotTableColumn, {
      context: {
        staticClass: 'foo',
        class: { bar: true },
        props: {
          sticky: 'left',
        },
      },
    });
    expect(stickyWrapper.classes('foo')).toBe(true);
    expect(stickyWrapper.classes('bar')).toBe(true);
    expect(stickyWrapper.classes('sticky-left')).toBe(true);
    expect(stickyWrapper.classes('sticky-right')).toBe(false);
  });

  it('does right sticky', () => {
    const stickyWrapper = mount(SlotTableColumn, {
      context: {
        staticClass: 'foo',
        class: { bar: true },
        props: {
          sticky: 'right',
        },
      },
    });
    expect(stickyWrapper.classes('foo')).toBe(true);
    expect(stickyWrapper.classes('bar')).toBe(true);
    expect(stickyWrapper.classes('sticky-right')).toBe(true);
    expect(stickyWrapper.classes('sticky-left')).toBe(false);
  });

  it('does no sticky', () => {
    const stickyWrapper = mount(SlotTableColumn, {
      context: {
        staticClass: 'foo',
        class: { bar: true },
        props: {
        },
      },
    });
    expect(stickyWrapper.classes('foo')).toBe(true);
    expect(stickyWrapper.classes('bar')).toBe(true);
    expect(stickyWrapper.classes('sticky-right')).toBe(false);
    expect(stickyWrapper.classes('sticky-left')).toBe(false);
  });
});
