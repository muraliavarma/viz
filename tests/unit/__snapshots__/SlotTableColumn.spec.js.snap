// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlotTableColumn.vue has the expected html structure 1`] = `
<div
  baz="quux"
  class="bar"
/>
`;
